tests/functional/test_plugins_adcm_hc.py

```diff
@@ -0,0 +1,179 @@
+        """Check that hc map is correct after adcm_hc action"""
+        actual_hc_map = {
+            (hc['host'], hc['service_name'], hc['component'])
+            for hc in adcm_client.cluster(name=CLUSTER_NAME).hostcomponent()
+        }
+        assert (
+            actual_hc_map == self.EXPECTED_HC_MAP
+        ), f"Host-Component map isn't the same as expected. Got: {actual_hc_map}. Expected: {self.EXPECTED_HC_MAP}"
+
+
+class TestPluginCanBreakConstraints:
+    """
+    Check that action that uses adcm_hc plugin can "break" component constraints
+    during execution if "in the end" constraints are satisfied
+    """
+
+    @pytest.fixture()
+    def cluster_with_service(self, sdk_client_fs: ADCMClient) -> Tuple[Cluster, Service]:
+        """Return cluster with service"""
+        bundle = sdk_client_fs.upload_from_fs(get_data_dir(__file__, 'clusters', 'constraints'))
+        cluster = bundle.cluster_create(CLUSTER_NAME)
+        test_service = cluster.service_add(name='test_service')
+        return cluster, test_service
+
+    @pytest.fixture()
+    def provider_with_hosts(self, sdk_client_fs: ADCMClient) -> Tuple[Provider, Host, Host, Host]:
+        """Return provider and three hosts"""
+        bundle = sdk_client_fs.upload_from_fs(get_data_dir(__file__, 'provider'))
+        provider = bundle.provider_create(PROVIDER_NAME)
+        hosts = tuple((provider.host_create(f'{prefix}-host') for prefix in ('first', 'second', 'third')))
+        return provider, *hosts
+
+    @pytest.fixture()
+    def set_hc_map(self, cluster_with_service, provider_with_hosts):
+        """Set HC map for test"""
+        cluster, test_service = cluster_with_service
+        _, first_host, second_host, third_host = provider_with_hosts
+        cluster.host_add(first_host)
+        cluster.host_add(second_host)
+        cluster.host_add(third_host)
+        return cluster.hostcomponent_set(
+            (first_host, test_service.component(name='test_component')),
+            (second_host, test_service.component(name='test_component')),
+        )
+
+    @pytest.mark.usefixtures("set_hc_map")
+    def test_correct_hc_change(self, cluster_with_service: Tuple[Cluster, Service]):
+        """
+        Check that correct "break" of HC constraints works.
+        Correct "break" is when constraints are satisfied after the action's end
+        but was "broken" during action execution.
+        """
+        cluster, _ = cluster_with_service
+        run_cluster_action_and_assert_result(cluster, 'correct_hc_map_change')
+
+    @pytest.mark.usefixtures("set_hc_map")
+    def test_incorrect_hc_change(self, cluster_with_service: Tuple[Cluster, Service]):
+        """
+        Check that incorrect "break" of HC constraints if forbidden.
+        Incorrect "break" is when constraints are "broken" during action's execution
+        and wasn't restored after all operations.
+        """
+        error = 'COMPONENT_CONSTRAINT_ERROR'
```

Comment:
mb put it in tests/library/errorcodes.py or somewhere with common consts?

Reply:
It isn't an error code, it's just a message in the job log, so I don't think it fits the case with errorcodes